### PR TITLE
Add setRequestDecoder to ResponseEncoder interface

### PR DESCRIPTION
**Pull request by pksohn**
Commit Message: Add setRequestDecoder to ResponseEncoder interface
Additional Description: This change was introduced in #23648 but was moved into a separate PR given its more general scope.
Risk Level: Low
Testing: No tests added.
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

File: envoy/http/codec.h [hunk 1]
```diff
@@ -165,6 +166,12 @@ class ResponseEncoder : public virtual StreamEncoder {
    * error.
    */
   virtual bool streamErrorOnInvalidHttpMessage() const PURE;
+
+  /**
+   * Set a new request decoder for this ResponseEncoder.
+   * @param decoder new request decoder.
```
Comment (alyssawilk) [inline envoy/http/codec.h:172]: think it's worth a comment on why this is a useful association?

File: source/common/http/http1/codec_impl.h [hunk 2]
```diff
@@ -156,6 +156,8 @@ class ResponseEncoderImpl : public StreamEncoderImpl, public ResponseEncoder {
     return stream_error_on_invalid_http_message_;
   }
 
+  void setRequestDecoder(Http::RequestDecoder& /*decoder*/) override {}
```
Comment (alyssawilk) [inline source/common/http/http1/codec_impl.h:159]: or maybe why HTTP/1.1 doesn't need this

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo one of two explanatory comments :-)

File: envoy/http/codec.h (hunk 1 above)
Comment (pksohn) [reply envoy/http/codec.h:174]: added, ptal!

File: source/common/http/http1/codec_impl.h (hunk 2 above)
Comment (pksohn) [reply source/common/http/http1/codec_impl.h:160]: Added. H/1 will need more work to actually reset the decoder on internal redirect, but it seems like we don't need to tackle that now?

File: source/common/http/conn_manager_impl.cc [hunk 3]
```diff
@@ -1767,6 +1764,12 @@ void ConnectionManagerImpl::ActiveStream::recreateStream(
   connection_manager_.doEndStream(*this, /*check_for_deferred_close*/ false);
 
   RequestDecoder& new_stream = connection_manager_.newStream(*response_encoder, true);
+  // Set the new RequestDecoder on the ResponseEncoder. Even though all of the decoder callbacks
+  // have already been called at this point, the encoder still needs the new decoder for deferred
+  // logging in some cases.
+  // This doesn't currently work for HTTP/1 as the H/1 ResponseEncoder doesn't hold the active
+  // stream's pointer to the RequestDecoder.
```
Comment (danzh2010) [inline source/common/http/conn_manager_impl.cc:1771]: For my own knowledge, how does H1 codec decode request headers in that case?

File: source/common/http/http1/codec_impl.h (hunk 2 above)
Comment (danzh2010) [reply source/common/http/http1/codec_impl.h:160]: We will need to make it work for H1 to defer logging I guess? Can you add a TODO?

File: source/common/http/conn_manager_impl.cc (hunk 3 above)
Comment (pksohn) [reply source/common/http/conn_manager_impl.cc:1771]: H1 has an [ActiveStream](https://github.com/envoyproxy/envoy/blob/28ac8d08d71b41544969d23415247dc53ce919c1/source/common/http/http1/codec_impl.h#L465-L466) that has a handle on both a ResponseEncoder and RequestDecoder. So e.g. on receiving headers the connection goes through the active request to get to the decoder, like [here](https://github.com/envoyproxy/envoy/blob/28ac8d08d71b41544969d23415247dc53ce919c1/source/common/http/http1/codec_impl.cc). It seems like the H/2 and H/3 codecs have the decoding path go through the response encoder -> request decoder partly to simplify multiplexing, so it makes sense that H/1 (which doesn't have streams) is organized a bit differently. I could be wrong about that last part.

**Review by alyssawilk — APPROVED** [review]

File: source/common/http/conn_manager_impl.cc (hunk 3 above)
Comment (danzh2010) [reply source/common/http/conn_manager_impl.cc:1771]: Get it! Thanks for the link!

**Review by danzh2010 — APPROVED** [review]
LGTM. Please update the PR description to reflect the recreateStream() change.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**